File: Stemmer/src/main/java/ph/bohol/util/stemmer/Stemmer.java
package ph.bohol.util.stemmer;

import java.util.HashMap;
import java.util.HashSet;
import java.util.LinkedList;
import java.util.Map;
import java.util.Set;
import java.util.Vector;

public class Stemmer {
    private String language;
    private boolean compiled = false;
    private final Map<String, String> constants = new HashMap<String, String>();
    private final Vector<AffixGroup> groups = new Vector<AffixGroup>();
    private RootWordProvider rootWordProvider = null;

    final void addGroup(final AffixGroup group) {
        groups.addElement(group);
        compiled = false;
    }

    final void compile() {
        for (AffixGroup group : groups) {
            group.compile(constants);
        }
        compiled = true;
    }

    final void addConstant(final String key, final String value) {
        constants.put(key, value);
    }

    public final LinkedList<Derivation> findDerivations(final String word) {
        if (!compiled) {
            compile();
        }
        Set<String> roots = new HashSet<String>();
        return innerFindDerivations(word, roots, 0);
    }


    private LinkedList<Derivation> innerFindDerivations(final String word, final Set<String> roots, final int level) {
        LinkedList<Derivation> derivations = null;
        if (groups.size() <= level) {
            derivations = new LinkedList<Derivation>();
            if (!roots.contains(word) && isRootWord(word)) {
                Derivation derivation = new Derivation(word);
                derivations.add(derivation);
                roots.add(word);
            }
            return derivations;
        } else {
            derivations = innerFindDerivations(word, roots, level + 1);
        }

        AffixGroup group = groups.get(level);
        LinkedList<Affix> affixes = group.getAffixes();
        for (Affix affix : affixes) {
            LinkedList<String> rootCandidates = affix.rootCandidates(word);
            for (String root : rootCandidates) {
                if (!roots.contains(root) && isRootWord(root)) {
                    Derivation derivation = new Derivation(root);
                    derivation.addAffix(affix);
                    derivations.add(derivation);
                    roots.add(root);
                }

                LinkedList<Derivation> innerDerivations = innerFindDerivations(root, roots, level + 1);

                // Copy the found derivations to the result list with the current affix as additional affix:
                for (Derivation derivation : innerDerivations) {
                    derivation.addAffix(affix);
                    derivations.add(derivation);
                }
            }
        }
        return derivations;
    }

    private boolean isRootWord(final String word) {
        return (rootWordProvider == null || rootWordProvider.isRootWord(word));
    }

    public final String getLanguage() {
        return language;
    }

    final void setLanguage(final String newLanguage) {
        this.language = newLanguage;
    }

    public final String toString() {
        String result = "<stemmer language='" + language + "'>";

        for (String key : constants.keySet()) {
            String value = constants.get(key);
            result += "\n<constant name='" + key + "' value='" + value + "'/>";
        }

        for (AffixGroup group : groups) {
            result += group.toString();
        }

        return result + "</stemmer>\n";
    }

    public final RootWordProvider getRootProvider() {
        return rootWordProvider;
    }

    public final void setRootProvider(final RootWordProvider rootProvider) {
        this.rootWordProvider = rootProvider;
    }
}

File: src/stemmer/Stemmer.ts
import {AffixGroup, compileGroup, toXmlGroup as groupToXml} from "./AffixGroup";
import {Derivation} from "./Derivation";
import {Affix, rootCandidates} from "./Affix";
import {RootWordProvider, isRootWord} from "./RootWordProvider";

export interface Stemmer {
    language: string;
    compiled: boolean;
    constants: {[x: string]: string};
    groups: AffixGroup[];
}
export function addGroup(stemmer: Stemmer, group: AffixGroup): void {
    stemmer.groups.push(group);
    stemmer.compiled = false;
}

export function compileStemmer(stemmer: Stemmer) {
    for (const group of stemmer.groups) compileGroup(group, stemmer.constants);
    stemmer.compiled = true;
}

export function addConstant(stemmer: Stemmer, key: string, value: string) {
    if (!!stemmer.constants[key] && value !== stemmer.constants[key])
        throw new Error(key + " was already set with " + stemmer.constants[key]
            + ", tried to set it with " + value);
    stemmer.constants[key] = value;
}

export function findDerivations(rootWordProvider: RootWordProvider, stemmer: Stemmer, word: string): Derivation[] {
    if (!stemmer.compiled) compileStemmer(stemmer);
    return innerFindDerivations(rootWordProvider, stemmer, word, {}, 0);
}


export function innerFindDerivations(rootWordProvider: RootWordProvider, stemmer: Stemmer, word: string, handledRoots: {[x: string]: boolean}, level: number): Derivation[] {
    let derivations: Derivation[];
    if (stemmer.groups.length <= level) {
        derivations = [];
        if (!handledRoots[word] && isRootWord(rootWordProvider, word)) {
            const affixes: Affix[] = [];
            const derivation = {root: word, affixes};
            derivations.push(derivation);
            handledRoots[word] = true;
        }
        return derivations;
    } else derivations = innerFindDerivations(rootWordProvider, stemmer, word, handledRoots, level + 1);

    const group: AffixGroup = stemmer.groups[level];
    const affixes: Affix[] = group.affixes;
    for (const affix of affixes) {
        const rootCandidatez: string[] = rootCandidates(word, affix.patterns);
        for (const root of rootCandidatez) {
            if (!handledRoots[root] && isRootWord(rootWordProvider, root)) {
                const affixes: Affix[] = [];
                const derivation = {root, affixes};
                derivation.affixes.push(affix);
                derivations.push(derivation);
                handledRoots[root] = true;
            }

            const innerDerivations = innerFindDerivations(rootWordProvider, stemmer, root, handledRoots, level + 1);

            // Copy the found derivations to the result list with the current affix as additional affix:
            for (const derivation of innerDerivations) {
                derivation.affixes.push(affix);
                derivations.push(derivation);
            }
        }
    }
    return derivations;
}


export function toString(stemmer: Stemmer): string {
    const result = ["<stemmer language='" + stemmer.language + "'>"];

    for (const key in stemmer.constants) if (stemmer.constants.hasOwnProperty(key)) {
        const value = stemmer.constants[key];
        result.push("\n<constant name='" + key + "' value='" + value + "'/>");
    }
    for (const group of stemmer.groups)  result.push(groupToXml(group));
    result.push("</stemmer>\n");
    return result.join("");
}
